### PR TITLE
feat 1621 Add validation rules for tax tariffs

### PR DESCRIPTION
**Pull request by ShaBloze**
## Description

This PR is about adding validation rules for tax tariffs.
Only System Operator (EZ) is allowed to submit requests concerning tax tariffs.

## References

* #1621

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRule.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using GreenEnergyHub.Charges.Domain.Charges;
+using GreenEnergyHub.Charges.Domain.Dtos.SharedDtos;
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands.Validation.BusinessValidation.ValidationRules
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRule.cs:19]: As per the tech note in the story, this rule should be enforced for both `ChargeInformation` and `ChargePrice`, hence it should probably be placed in some shared namespace.

Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRule.cs:19]: This rule is related to `ChargePrice` and should be placed in a corresponding namespace.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs [hunk 2]
```diff
@@ -59,5 +59,7 @@ public enum ValidationRuleIdentifier
         TaxIndicatorMustBeFalseForFee = 48, // VR921 / D14
         ChargeOperationIdLengthValidation = 49, // VR922 / E86
         ChargeOwnerMustMatchSender = 50, // VR513 / E0I
+        ChargeTypeTariffTaxIndicator = 51, // VR508 / E0I
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs:62]:
```suggestion
        ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperator = 51, // VR508 / E0I
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs [hunk 3]
```diff
@@ -59,5 +59,7 @@ public enum ValidationRuleIdentifier
...
+        UpdateTaxTariffOnlyBySystemOperator = 52, // VR508 / E0I
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs:63]:
```suggestion
        UpdateTaxTariffOnlyAllowedBySystemOperator = 52, // VR508 / E0I
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePriceEventHandlerDeprecatedTests.cs [hunk 4]
```diff
@@ -225,6 +229,65 @@ public class ChargePriceEventHandlerDeprecatedTests
...
+                    It.IsAny<DocumentDto>()))
+                .Callback<IReadOnlyCollection<ChargeInformationOperationDto>, DocumentDto>((_, _) => accepted++);
+            var rejectedRules = new List<IValidationRuleContainer>();
+            receiptService
+                .Setup(s => s.RejectInvalidOperationsAsync(
+                    It.IsAny<IReadOnlyCollection<ChargeInformationOperationDto>>(),
+                    It.IsAny<DocumentDto>(),
+                    It.IsAny<IList<IValidationRuleContainer>>()))
+                .Callback<IReadOnlyCollection<ChargeInformationOperationDto>, DocumentDto, IList<IValidationRuleContainer>>(
+                    (_, _, s) => rejectedRules.AddRange(s));
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePriceEventHandlerDeprecatedTests.cs:270]: Move to a `SetupReceiptServiceMock` to enhance readability of the test method and enable code reuse across `Task HandleAsync_WhenChargeUpdateHasStartDateAfterStopDate_RejectCurrentAndAllSubsequentOperations`/`HandleAsync_WhenTaxTariffButNotSystemOperator_RejectCurrentAndAllSubsequentOperations`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeBuilder.cs [hunk 5]
```diff
@@ -135,7 +143,8 @@ public Charge Build()
                     _points.Min(p => p.Time),
                     _points.Max(p => p.Time),
                     _points,
-                    operationId);
+                    operationId,
+                    MarketParticipantRole.SystemOperator);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeBuilder.cs:147]: Could be, that we would rather use another role as default? @prtandrup?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Charges/ChargeTests.cs [hunk 6]
```diff
@@ -531,6 +534,42 @@ public void UpdatePrices_WhenNoPointsExistBetweenStartAndStopDate_PointsUpdated(
...
+        {
+            // Arrange
+            var points = new List<Point>
+            {
+                new(1.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(0)),
+                new(2.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(1)),
+                new(3.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(2)),
+                new(4.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(3)),
+                new(5.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(4)),
+            };
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Charges/ChargeTests.cs:548]: Could we do fine with just one existing point?
```suggestion
            var points = new List<Point> { new(1.00m, InstantHelper.GetTodayPlusDaysAtMidnightUtc(0)) };
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeBuilder.cs (hunk 5 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Builders/Command/ChargeBuilder.cs:147]: Perhaps GridAccessProvider would be slightly better, although system operator is indeed valid. If necessary we could WithMarketParticipant method.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        [Theory]
+        [InlineAutoMoqData(ChargeType.Fee)]
+        [InlineAutoMoqData(ChargeType.Subscription)]
+        [InlineAutoMoqData(ChargeType.Unknown)]
+        public void ChargeTypeTariffTaxIndicatorAuthorizeRule_WhenChargeTypeNotTariff_ReturnsTrue(
+            ChargeType chargeType)
+        {
+            // Arrange
+            // Act
+            var sut = new UpdateTaxTariffOnlyBySystemOperatorRule(chargeType, MarketParticipantRole.SystemOperator, false);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs:38]: The test will always return true regardless of the `chargeType`provided due to the use of the `MarketParticipantRole.SystemOperator` role. Charge to another role, and the test will be fine.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+            // Assert
+            sut.IsValid.Should().Be(true);
+        }
+
+        [Theory]
+        [InlineAutoMoqData(MarketParticipantRole.GridAccessProvider, true, false)]
+        [InlineAutoMoqData(MarketParticipantRole.GridAccessProvider, false, true)]
+        [InlineAutoMoqData(MarketParticipantRole.SystemOperator, true, true)]
+        [InlineAutoMoqData(MarketParticipantRole.SystemOperator, false, true)]
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs:48]: `MarketParticipantRole` can have many other values, and despite being quite static the `MarketParticipantRole` enum could be changed to contain more or fewer roles, so I suggest we build the test to embrace change by testing all roles in the current enum: 
```
foreach (var senderRole in Enum.GetValues<MarketParticipantRole>())
{
    // Arrange
    var expectedResult = senderRole == MarketParticipantRole.SystemOperator

    // Act
    var sut = new UpdateTaxTariffOnlyBySystemOperatorRule(ChargeType.Tariff, senderRole, taxIndicator);

    // Assert
    sut.IsValid.Should().Be(expectedResult);
}
```
with taxIndicator still provided by _InlineAutoMoqData_

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorValidationRuleTests.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        [InlineAutoMoqData(ChargeType.Subscription)]
+        [InlineAutoMoqData(ChargeType.Unknown)]
+        public void ChargeTypeTariffTaxIndicatorValidationRule_WhenChargeTypeNotTariff_ReturnsTrue(
+            ChargeType chargeType,
+            ChargeInformationOperationDtoBuilder chargeInformationOperationDtoBuilder,
+            MarketParticipantDtoBuilder marketParticipantDtoBuilder)
+        {
+            // Arrange
+            var senderMarketParticipant = marketParticipantDtoBuilder
+                .WithMarketParticipantRole(MarketParticipantRole.SystemOperator)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorValidationRuleTests.cs:42]: The test will always return true regardless of the `chargeType`provided due to the use of the `MarketParticipantRole.SystemOperator` role. Charge to another role, and the test will be fine.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorValidationRuleTests.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            sut.IsValid.Should().Be(true);
+        }
+
+        [Theory]
+        [InlineAutoMoqData(MarketParticipantRole.GridAccessProvider, TaxIndicator.Tax, false)]
+        [InlineAutoMoqData(MarketParticipantRole.GridAccessProvider, TaxIndicator.NoTax, true)]
+        [InlineAutoMoqData(MarketParticipantRole.GridAccessProvider, TaxIndicator.Unknown, true)]
+        [InlineAutoMoqData(MarketParticipantRole.SystemOperator, TaxIndicator.Tax, true)]
+        [InlineAutoMoqData(MarketParticipantRole.SystemOperator, TaxIndicator.NoTax, true)]
+        [InlineAutoMoqData(MarketParticipantRole.SystemOperator, TaxIndicator.Unknown, true)]
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorValidationRuleTests.cs:61]: Se my comment to the similar test in `UpdateTaxTariffOnlyBySystemOperatorRuleTests`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            // Arrange
+            // Act
+            var sut = new UpdateTaxTariffOnlyBySystemOperatorRule(ChargeType.Tariff, senderRole, taxIndicator);
+
+            // Assert
+            sut.IsValid.Should().Be(expectedResult);
+        }
+
+        [Fact]
+        public void ValidationRuleIdentifier_ShouldBe_EqualTo()
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs:63]:
```suggestion
        public void ValidationRuleIdentifier_ShouldBe_UpdateTaxTariffOnlyBySystemOperator()
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Charges/ChargeTests.cs (hunk 6 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Charges/ChargeTests.cs:548]: Yes that would be fine. I have changed it to only one.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePriceEventHandlerDeprecatedTests.cs (hunk 4 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePriceEventHandlerDeprecatedTests.cs]: Merged main into this branch, which meant that this test has been removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs (hunk 8 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyBySystemOperatorRuleTests.cs]: Changed the test to test all roles. I removed the taxIndicator and replaced it with _true_.

I added a new test to ensure that when TaxIndicator is false, then IsValid is true.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands.Validation.InputValidation.ValidationRules
+{
+    public class ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule : IValidationRule
+    {
+        private readonly ChargeInformationOperationDto _chargeInformationOperationDto;
+        private readonly MarketParticipantDto _senderMarketParticipantDto;
+
+        public ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule(ChargeInformationOperationDto chargeInformationOperationDto, MarketParticipantDto senderMarketParticipantDto)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule.cs:26]:
```suggestion
        public ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule(
            ChargeInformationOperationDto chargeInformationOperationDto,
            MarketParticipantDto senderMarketParticipantDto)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            _chargeInformationOperationDto = chargeInformationOperationDto;
+            _senderMarketParticipantDto = senderMarketParticipantDto;
+        }
+
+        public bool IsValid => _chargeInformationOperationDto.ChargeType is not ChargeType.Tariff ||
+                               _chargeInformationOperationDto.TaxIndicator is not TaxIndicator.Tax ||
+                               _senderMarketParticipantDto.BusinessProcessRole is MarketParticipantRole.SystemOperator;
+
+        public ValidationRuleIdentifier ValidationRuleIdentifier => ValidationRuleIdentifier.ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperator;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule.cs:36]:
```suggestion
        public ValidationRuleIdentifier ValidationRuleIdentifier =>
            ValidationRuleIdentifier.ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperator;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRule.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands.Validation.BusinessValidation.ValidationRules
+{
+    public class UpdateTaxTariffOnlyAllowedBySystemOperatorRule : IValidationRule
+    {
+        private readonly ChargeType _chargeType;
+        private readonly MarketParticipantRole _businessProcessRole;
+        private readonly bool _taxIndicator;
+
+        public UpdateTaxTariffOnlyAllowedBySystemOperatorRule(ChargeType chargeType, MarketParticipantRole businessProcessRole, bool taxIndicator)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRule.cs:27]:
```suggestion
        public UpdateTaxTariffOnlyAllowedBySystemOperatorRule(
            ChargeType chargeType, MarketParticipantRole businessProcessRole, bool taxIndicator)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRule.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            _chargeType = chargeType;
+            _businessProcessRole = businessProcessRole;
+            _taxIndicator = taxIndicator;
+        }
+
+        public bool IsValid => _chargeType is not ChargeType.Tariff ||
+                               !_taxIndicator ||
+                               _businessProcessRole is MarketParticipantRole.SystemOperator;
+
+        public ValidationRuleIdentifier ValidationRuleIdentifier => ValidationRuleIdentifier.UpdateTaxTariffOnlyAllowedBySystemOperator;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRule.cs:38]:
```suggestion
        public ValidationRuleIdentifier ValidationRuleIdentifier =>
            ValidationRuleIdentifier.UpdateTaxTariffOnlyAllowedBySystemOperator;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRuleTests.cs [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+            // Act
+            var sut = new ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule(chargeOperation, senderMarketParticipant);
+
+            // Assert
+            sut.IsValid.Should().Be(true);
+        }
+
+        [Theory]
+        [AutoDomainData]
+        public void IsValid_WhenNotSystemOperator_IsTrue(
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRuleTests.cs:58]:
```suggestion
        public void IsValid_WhenNotSystemOperator_IsFalse(
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRuleTests.cs [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+                .Build();
+            var chargeOperation = chargeInformationOperationDtoBuilder
+                .WithChargeType(ChargeType.Tariff)
+                .Build();
+
+            // Act
+            var sut = new ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule(chargeOperation, senderMarketParticipant);
+
+            // Assert
+            sut.ValidationRuleIdentifier.Should().Be(ValidationRuleIdentifier.ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperator);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRuleTests.cs:122]:
```suggestion
            var sut = new ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperatorValidationRule(
                chargeOperation, senderMarketParticipant);

            // Assert
            sut.ValidationRuleIdentifier.Should()
                .Be(ValidationRuleIdentifier.ChargeTypeTariffTaxIndicatorOnlyAllowedBySystemOperator);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+{
+    [UnitTest]
+    public class UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests
+    {
+        [Theory]
+        [InlineAutoMoqData(ChargeType.Fee)]
+        [InlineAutoMoqData(ChargeType.Subscription)]
+        [InlineAutoMoqData(ChargeType.Unknown)]
+        public void IsValid_WhenChargeTypeNotTariff_IsTrue(
+            ChargeType chargeType)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs:35]:
```suggestion
        public void IsValid_WhenChargeTypeNotTariff_IsTrue(ChargeType chargeType)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace GreenEnergyHub.Charges.Tests.Domain.Dtos.ChargePriceCommands.Validation.BusinessRules.ValidationRules
+{
+    [UnitTest]
+    public class UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests
+    {
+        [Theory]
+        [InlineAutoMoqData(ChargeType.Fee)]
+        [InlineAutoMoqData(ChargeType.Subscription)]
+        [InlineAutoMoqData(ChargeType.Unknown)]
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs:33]: If we change the test to loop all values of `ChargeType`, then the test would also be valid if another type of charge is added. I know it is unlikely going to happen :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+            // Arrange
+            // Act
+            var sut = new UpdateTaxTariffOnlyAllowedBySystemOperatorRule(chargeType, MarketParticipantRole.GridAccessProvider, false);
+
+            // Assert
+            sut.IsValid.Should().Be(true);
+        }
+
+        [Fact]
+        public void IsValid_WhenNotSystemOperator_IsTrue()
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs:46]:
```suggestion
        public void IsValid_WhenNotSystemOperator_IsFalse()
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        public void ValidationRuleIdentifier_ShouldBeUpdateTaxTariffOnlyAllowedBySystemOperator()
+        {
+            // Arrange
+            // Act
+            var sut = new UpdateTaxTariffOnlyAllowedBySystemOperatorRule(ChargeType.Tariff, MarketParticipantRole.SystemOperator, true);
+
+            // Assert
+            sut.ValidationRuleIdentifier.Should().Be(ValidationRuleIdentifier.UpdateTaxTariffOnlyAllowedBySystemOperator);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs:83]:
```suggestion
            var sut = new UpdateTaxTariffOnlyAllowedBySystemOperatorRule(
                ChargeType.Tariff, MarketParticipantRole.SystemOperator, true);

            // Assert
            sut.ValidationRuleIdentifier.Should()
                .Be(ValidationRuleIdentifier.UpdateTaxTariffOnlyAllowedBySystemOperator);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs (hunk 19 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/BusinessRules/ValidationRules/UpdateTaxTariffOnlyAllowedBySystemOperatorRuleTests.cs:33]: Changed the test to loop through all ChargeTypes, although I had to exclude Tariff.

**Review by HenrikSommer — APPROVED** [review]

**Pull request merged**